In [1]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.


# AI Lesson 04a Tasks: SQL for Machine Learning

**Course:** Applications of Artificial Intelligence  
**Lesson:** SQL Introduction with NBA Dataset  

---

## Learning Objectives

By completing this task, you will:
- Write SQL queries to extract data for machine learning
- Use SELECT, WHERE, ORDER BY, and LIMIT effectively
- Filter data to create better training datasets
- Load SQL results into pandas DataFrames
- Export data to Excel for analysis
- Understand how SQL fits into the ML pipeline

---

## The ML Context

**Why are we doing this?**

In real-world machine learning:
- Data lives in **databases**, not CSV files
- You need to **filter** data before loading (millions of rows!)
- SQL lets you **select specific features** for your model
- SQL helps you **create training datasets** efficiently
- You export data to Excel for visualization and analysis

**Today's scenario:**
You're building an ML model to predict NBA team performance. First, you need to extract the right data from the database using SQL, then export it to Excel for analysis.

---

## Setup: Connect to Database

In [1]:
# Import necessary libraries
import pandas as pd
import sqlite3

# Connect to the NBA database
conn = sqlite3.connect('nba_5seasons.db')

print("✅ Connected to NBA database")
print("Ready to write SQL queries!")

✅ Connected to NBA database
Ready to write SQL queries!


---
## Part 1: Understanding the Database Structure

Before we can write good queries, we need to know what tables and columns we have.

### Task 1.1: List All Tables

Write a query to see what tables exist in our database.

**Hint:** Use the `sqlite_master` table and filter for type='table'

In [2]:
# TODO: Write a query to list all tables
query = """
SELECT *
FROM sqlite_master
WHERE type = 'table'
"""

tables = pd.read_sql(query, conn)
print("Tables in the database:")
display(tables)

Tables in the database:


type                 name             tbl_name  rootpage  \
0  table                teams                teams         2   
1  table              players              players         3   
2  table      team_game_stats      team_game_stats         4   
3  table  player_season_stats  player_season_stats         6   

                                                 sql  
0  CREATE TABLE teams (\n      team_id INTEGER PR...  
1  CREATE TABLE players (\n      player_id INTEGE...  
2  CREATE TABLE team_game_stats (\n      season T...  
3  CREATE TABLE player_season_stats (\n      seas...

**Q1: How many tables are in the database?**

A: 4

### Task 1.2: Explore the teams Table

Look at the first 10 rows of the `teams` table.

**Hint:** Use SELECT * and LIMIT

In [3]:
# TODO: Write a query to see the first 10 teams
query = """
SELECT *
FROM teams
LIMIT 10
"""

teams_sample = pd.read_sql(query, conn)
print("First 10 teams:")
display(teams_sample)

First 10 teams:


team_id              full_name abbreviation   nickname           city  \
0  1610612737          Atlanta Hawks          ATL      Hawks        Atlanta   
1  1610612738         Boston Celtics          BOS    Celtics         Boston   
2  1610612739    Cleveland Cavaliers          CLE  Cavaliers      Cleveland   
3  1610612740   New Orleans Pelicans          NOP   Pelicans    New Orleans   
4  1610612741          Chicago Bulls          CHI      Bulls        Chicago   
5  1610612742       Dallas Mavericks          DAL  Mavericks         Dallas   
6  1610612743         Denver Nuggets          DEN    Nuggets         Denver   
7  1610612744  Golden State Warriors          GSW   Warriors  San Francisco   
8  1610612745        Houston Rockets          HOU    Rockets        Houston   
9  1610612746   Los Angeles Clippers          LAC   Clippers    Los Angeles   

           state  year_founded  
0        Georgia          1949  
1  Massachusetts          1946  
2           Ohio          1970  
3      Louisiana          2002  
4       Illinois          1966  
5          Texas          1980  
6       Colorado          1976  
7     California          1946  
8          Texas          1967  
9     California          1970

**Q2: What columns are in the teams table?**
A: team_id, full_name, abbreviation, mickname, city, state, and year founded.

**Q3: How would you find your favorite team in this table?**

A: Via name?

---
## Part 2: Basic SELECT Queries

Let's practice selecting specific columns - this is how you choose **features** for ML!

### Task 2.1: Select Specific Columns

Get just the team names and cities (the features we care about).

**Hint:** SELECT column1, column2 FROM table

In [4]:
# TODO: Select only full_name and city columns
query = """
SELECT full_name, city
FROM teams
"""

team_names = pd.read_sql(query, conn)
print("Team names and cities:")
display(team_names)

Team names and cities:


full_name           city
0            Atlanta Hawks        Atlanta
1           Boston Celtics         Boston
2      Cleveland Cavaliers      Cleveland
3     New Orleans Pelicans    New Orleans
4            Chicago Bulls        Chicago
5         Dallas Mavericks         Dallas
6           Denver Nuggets         Denver
7    Golden State Warriors  San Francisco
8          Houston Rockets        Houston
9     Los Angeles Clippers    Los Angeles
10      Los Angeles Lakers    Los Angeles
11              Miami Heat          Miami
12         Milwaukee Bucks      Milwaukee
13  Minnesota Timberwolves      Minnesota
14           Brooklyn Nets       Brooklyn
15         New York Knicks       New York
16           Orlando Magic        Orlando
17          Indiana Pacers        Indiana
18      Philadelphia 76ers   Philadelphia
19            Phoenix Suns        Phoenix
20  Portland Trail Blazers       Portland
21        Sacramento Kings     Sacramento
22       San Antonio Spurs    San Antonio
23   Oklahoma City Thunder  Oklahoma City
24         Toronto Raptors        Toronto
25               Utah Jazz           Utah
26       Memphis Grizzlies        Memphis
27      Washington Wizards     Washington
28         Detroit Pistons        Detroit
29       Charlotte Hornets      Charlotte

**Q4: Why might you want to select only specific columns instead of using SELECT * ?**

A: Because there is alot of not-needed information that makes the data seem cluttered.

### Task 2.2: Teams by State

Let's see which states have NBA teams. Get the team name, city, and state, ordered by state.

**Hint:** ORDER BY can take multiple columns

In [5]:
# TODO: Show team names, cities, and states, ordered by state then name
query = """
SELECT full_name, city, state
FROM teams
ORDER BY state, full_name
"""

teams_by_state = pd.read_sql(query, conn)
print("All teams by state:")
display(teams_by_state)

All teams by state:


full_name           city                 state
0             Phoenix Suns        Phoenix               Arizona
1    Golden State Warriors  San Francisco            California
2     Los Angeles Clippers    Los Angeles            California
3       Los Angeles Lakers    Los Angeles            California
4         Sacramento Kings     Sacramento            California
5           Denver Nuggets         Denver              Colorado
6       Washington Wizards     Washington  District of Columbia
7               Miami Heat          Miami               Florida
8            Orlando Magic        Orlando               Florida
9            Atlanta Hawks        Atlanta               Georgia
10           Chicago Bulls        Chicago              Illinois
11          Indiana Pacers        Indiana               Indiana
12    New Orleans Pelicans    New Orleans             Louisiana
13          Boston Celtics         Boston         Massachusetts
14         Detroit Pistons        Detroit              Michigan
15  Minnesota Timberwolves      Minnesota             Minnesota
16           Brooklyn Nets       Brooklyn              New York
17         New York Knicks       New York              New York
18       Charlotte Hornets      Charlotte        North Carolina
19     Cleveland Cavaliers      Cleveland                  Ohio
20   Oklahoma City Thunder  Oklahoma City              Oklahoma
21         Toronto Raptors        Toronto               Ontario
22  Portland Trail Blazers       Portland                Oregon
23      Philadelphia 76ers   Philadelphia          Pennsylvania
24       Memphis Grizzlies        Memphis             Tennessee
25        Dallas Mavericks         Dallas                 Texas
26         Houston Rockets        Houston                 Texas
27       San Antonio Spurs    San Antonio                 Texas
28               Utah Jazz           Utah                  Utah
29         Milwaukee Bucks      Milwaukee             Wisconsin

**Q5: Which state has the most NBA teams?**

A: California

---
## Part 3: WHERE Clause - Filtering for ML

In ML, you often want **specific subsets** of data. WHERE is how you filter!

### Task 3.1: Filter by Season

Get all games from the 2021-22 season. (If you're training a model on recent data, you'd start here!)

**Hint:** WHERE column = 'value' (use single quotes for text!)

In [7]:
# TODO: Filter for only 2021-22 season games
query = """
SELECT *
FROM team_game_stats
WHERE season = '2021-22'
LIMIT 10
"""

recent_games = pd.read_sql(query, conn)
print("Sample of 2021-22 season games:")
display(recent_games)

Sample of 2021-22 season games:


season     game_id     team_id   game_date      matchup wl  pts  fgm  fga  \
0  2021-22  0022100001  1610612749  2021-10-19  MIL vs. BKN  W  127   48  105   
1  2021-22  0022100001  1610612751  2021-10-19    BKN @ MIL  L  104   37   84   
2  2021-22  0022100002  1610612744  2021-10-19    GSW @ LAL  W  121   41   93   
3  2021-22  0022100002  1610612747  2021-10-19  LAL vs. GSW  L  114   45   95   
4  2021-22  0022100003  1610612754  2021-10-20    IND @ CHA  L  122   42   90   
5  2021-22  0022100003  1610612766  2021-10-20  CHA vs. IND  W  123   46  107   
6  2021-22  0022100004  1610612741  2021-10-20    CHI @ DET  W   94   37   86   
7  2021-22  0022100004  1610612765  2021-10-20  DET vs. CHI  L   88   36   90   
8  2021-22  0022100005  1610612738  2021-10-20    BOS @ NYK  L  134   48  117   
9  2021-22  0022100005  1610612752  2021-10-20  NYK vs. BOS  W  138   51  105   

   fg3m  ...  ftm  fta  oreb  dreb  reb  ast  stl  blk  tov  plus_minus  
0    17  ...   14   18    13    41   54   25    8    9    8          23  
1    17  ...   13   23     5    39   44   19    3    9   13         -23  
2    14  ...   25   30     9    41   50   30    9    2   17           7  
3    15  ...    9   19     5    40   45   21    7    4   18          -7  
4    17  ...   21   24     8    43   51   29    2   10   17          -1  
5    13  ...   18   27    12    34   46   29    9    5    8           1  
6     7  ...   13   15     9    39   48   18    8    5   17           6  
7     6  ...   10   13    11    36   47   17    7    5   17          -6  
8    21  ...   17   23    15    41   56   34   13    9   18          -4  
9    17  ...   19   27     7    48   55   27    9   10   19           4  

[10 rows x 21 columns]

**Q6: Why use quotes around '2021-22'?**

A: Because it's a string, not an integer.

### Task 3.2: High-Scoring Games

Find all games where a team scored 120 or more points.

**ML Context:** This could be your positive class for a "high-scoring game predictor" model!

**Hint:** Use >= for "greater than or equal to"

In [8]:
# TODO: Find games with 120+ points, show top 20
query = """
SELECT season, game_date, team_id, matchup, pts, wl
FROM team_game_stats
WHERE pts >= 120
ORDER BY pts DESC
LIMIT 20
"""

high_scoring = pd.read_sql(query, conn)
print("Top 20 highest-scoring games:")
display(high_scoring)

Top 20 highest-scoring games:


season   game_date     team_id      matchup  pts wl
0   2022-23  2023-02-24  1610612758    SAC @ LAC  176  W
1   2022-23  2023-02-24  1610612746  LAC vs. SAC  175  L
2   2024-25  2025-03-27  1610612754    IND @ WAS  162  W
3   2021-22  2022-01-26  1610612766    CHA @ IND  158  W
4   2021-22  2022-02-25  1610612759    SAS @ WAS  157  W
5   2022-23  2023-04-09  1610612744    GSW @ POR  157  W
6   2023-24  2023-11-21  1610612754    IND @ ATL  157  W
7   2023-24  2024-04-14  1610612754  IND vs. ATL  157  W
8   2023-24  2023-11-01  1610612738  BOS vs. IND  155  W
9   2024-25  2024-12-26  1610612763  MEM vs. TOR  155  W
10  2023-24  2024-01-03  1610612762  UTA vs. DET  154  W
11  2021-22  2022-02-25  1610612764  WAS vs. SAS  153  L
12  2021-22  2022-04-01  1610612746    LAC @ MIL  153  W
13  2022-23  2022-11-15  1610612758  SAC vs. BKN  153  W
14  2022-23  2023-02-04  1610612760  OKC vs. HOU  153  W
15  2023-24  2024-01-23  1610612740  NOP vs. UTA  153  W
16  2024-25  2025-04-11  1610612748    MIA @ NOP  153  W
17  2021-22  2021-12-02  1610612763  MEM vs. OKC  152  W
18  2023-24  2023-11-06  1610612754  IND vs. SAS  152  W
19  2023-24  2023-11-21  1610612737  ATL vs. IND  152  L

**Q7: Looking at the results, what was the highest score?**

A: 176

**Q8: Did high-scoring games tend to be wins or losses?**

A: Wins

### 📊 Task 3.3: Export High-Scoring Games to Excel

Let's export ALL high-scoring games (120+ points) to Excel for analysis!

**You'll use this file later to practice Excel formulas and charts.**

In [9]:
# TODO: Get ALL high-scoring games (remove LIMIT) and export to Excel
query = """
SELECT season, game_date, team_id, matchup, pts, fgm, fga, fg3m, ftm, reb, ast, wl
FROM team_game_stats
WHERE pts >= 120
ORDER BY pts DESC
"""

high_scoring_all = pd.read_sql(query, conn)

# TODO: Export to Excel (fill in the blanks)
high_scoring_all.to_excel(
    "high_scoring_games.xlsx",
    index=False,
    sheet_name="High Scoring Games"
)

print(f"✅ Exported {len(high_scoring_all)} high-scoring games to 'high_scoring_games.xlsx'")
print("\nPreview:")
display(high_scoring_all.head())

✅ Exported 3427 high-scoring games to 'high_scoring_games.xlsx'

Preview:


season   game_date     team_id      matchup  pts  fgm  fga  fg3m  ftm  \
0  2022-23  2023-02-24  1610612758    SAC @ LAC  176   65  111    18   28   
1  2022-23  2023-02-24  1610612746  LAC vs. SAC  175   59   98    26   31   
2  2024-25  2025-03-27  1610612754    IND @ WAS  162   59   92    27   17   
3  2021-22  2022-01-26  1610612766    CHA @ IND  158   54   93    24   26   
4  2021-22  2022-02-25  1610612759    SAS @ WAS  157   57  105     9   34   

   reb  ast wl  
0   43   33  W  
1   42   39  L  
2   50   48  W  
3   36   37  W  
4   50   35  W

**Q9: How many high-scoring games were exported?**

A: 5 high-scoring games were exported.

---
## Part 4: ORDER BY and Multiple Conditions

### Task 4.1: Wins with Good Offense

Find all wins where the team scored 115+ points.

**Hint:** Combine wl = 'W' AND pts >= 115

In [10]:
# TODO: Find wins with 115+ points from 2021-22 season
query = """
SELECT game_date, team_id, matchup, pts, wl
FROM team_game_stats
WHERE season = '2021-22'
  AND wl = 'W'
  AND pts >= 115
ORDER BY pts DESC
LIMIT 25
"""

dominant_wins = pd.read_sql(query, conn)
print("Dominant wins (115+ points):")
display(dominant_wins)

Dominant wins (115+ points):


game_date     team_id      matchup  pts wl
0   2022-01-26  1610612766    CHA @ IND  158  W
1   2022-02-25  1610612759    SAS @ WAS  157  W
2   2022-04-01  1610612746    LAC @ MIL  153  W
3   2021-12-02  1610612763  MEM vs. OKC  152  W
4   2022-03-15  1610612751    BKN @ ORL  150  W
5   2022-03-14  1610612750    MIN @ SAS  149  W
6   2021-11-27  1610612745  HOU vs. CHA  146  W
7   2022-04-10  1610612747    LAL @ DEN  146  W
8   2021-12-04  1610612738    BOS @ POR  145  W
9   2021-12-26  1610612739  CLE vs. TOR  144  W
10  2021-12-26  1610612759  SAS vs. DET  144  W
11  2022-04-02  1610612755  PHI vs. CHA  144  W
12  2022-04-03  1610612738  BOS vs. WAS  144  W
13  2022-04-05  1610612748  MIA vs. CHA  144  W
14  2021-12-08  1610612758  SAC vs. ORL  142  W
15  2022-02-17  1610612746  LAC vs. HOU  142  W
16  2022-03-08  1610612749    MIL @ OKC  142  W
17  2022-03-11  1610612766    CHA @ NOP  142  W
18  2021-11-26  1610612758    SAC @ LAL  141  W
19  2022-01-09  1610612750    MIN @ HOU  141  W
20  2022-02-11  1610612766    CHA @ DET  141  W
21  2022-04-09  1610612763  MEM vs. NOP  141  W
22  2021-10-25  1610612738    BOS @ CHA  140  W
23  2021-11-05  1610612758  SAC vs. CHA  140  W
24  2022-01-05  1610612766  CHA vs. DET  140  W

**Q10: Why do we need quotes around 'W' but not around 115?**

A: Because 115 is an integer.

**Q11: What does the AND operator do in SQL?**

A: adds another condition to the elif.

### Task 4.2: Losses Despite Good Scoring

Find losses where the team still scored 110+ points. (Lost a shootout!)

**Hint:** Similar to above, but wl = 'L'

In [11]:
# TODO: Find losses with 110+ points (high-scoring losses)
query = """
SELECT game_date, team_id, matchup, pts, wl
FROM team_game_stats
WHERE season = '2021-22'
  AND wl = 'L'
  AND pts >= 115
ORDER BY pts DESC
LIMIT 25
"""

shootout_losses = pd.read_sql(query, conn)
print("High-scoring losses (110+ pts but still lost):")
display(shootout_losses)

High-scoring losses (110+ pts but still lost):


game_date     team_id      matchup  pts wl
0   2022-02-25  1610612764  WAS vs. SAS  153  L
1   2021-11-27  1610612766    CHA @ HOU  143  L
2   2022-04-10  1610612743  DEN vs. LAL  141  L
3   2022-03-14  1610612759  SAS vs. MIN  139  L
4   2021-11-26  1610612747  LAL vs. SAC  137  L
5   2021-10-20  1610612738    BOS @ NYK  134  L
6   2021-10-22  1610612754    IND @ WAS  134  L
7   2022-01-17  1610612754    IND @ LAC  133  L
8   2022-04-03  1610612745  HOU vs. MIN  132  L
9   2022-01-03  1610612737    ATL @ POR  131  L
10  2022-01-19  1610612758  SAC vs. DET  131  L
11  2022-03-28  1610612757  POR vs. OKC  131  L
12  2021-12-03  1610612738    BOS @ UTA  130  L
13  2021-12-08  1610612753    ORL @ SAC  130  L
14  2022-03-06  1610612740    NOP @ DEN  130  L
15  2022-03-09  1610612747    LAL @ HOU  130  L
16  2022-03-24  1610612743  DEN vs. PHX  130  L
17  2022-03-31  1610612746    LAC @ CHI  130  L
18  2022-04-01  1610612743  DEN vs. MIN  130  L
19  2021-10-25  1610612766  CHA vs. BOS  129  L
20  2021-12-14  1610612761    TOR @ BKN  129  L
21  2022-02-26  1610612759    SAS @ MIA  129  L
22  2022-01-19  1610612746    LAC @ DEN  128  L
23  2022-03-13  1610612754    IND @ ATL  128  L
24  2021-12-05  1610612737  ATL vs. CHA  127  L

**Q12: What does this tell you about offense vs defense in basketball?**

A: I have no idea what offense and defense means in terms of basketball.

---
## Part 5: Player Season Stats

Now let's look at player data - this is where ML gets interesting!

### Task 5.1: Explore Player Stats

Look at the structure of the player_season_stats table.

In [12]:
# TODO: Get first 10 rows from player_season_stats for 2021-22
query = """
SELECT *
FROM player_season_stats
WHERE season = '2021-22'
LIMIT 10
"""

player_stats = pd.read_sql(query, conn)
print("Sample player season stats:")
display(player_stats)

Sample player season stats:


season  player_id     team_id  gp          min     pts    reb    ast  \
0  2021-22       2207  1610612738   1     1.950000     2.0    0.0    0.0   
1  2021-22       2544  1610612747  56  2084.335000  1695.0  459.0  349.0   
2  2021-22       2546  1610612747  69  1793.190000   919.0  288.0   68.0   
3  2021-22       2617  1610612748  13    83.216667    32.0   25.0    4.0   
4  2021-22       2730  1610612747  60   970.723333   372.0  356.0   35.0   
5  2021-22       2738  1610612744  31   602.831667   124.0  100.0  114.0   
6  2021-22       2772  1610612747  24   462.303333    96.0   82.0   27.0   
7  2021-22     101108  1610612756  65  2138.958333   958.0  283.0  702.0   
8  2021-22     101139  1610612738   1     1.950000     0.0    0.0    0.0   
9  2021-22     101150  1610612737  56   798.876667   354.0   87.0  105.0   

     stl   blk    tov  fg_pct  fg3_pct  ft_pct  
0    0.0   0.0    0.0   1.000    0.000   0.000  
1   73.0  59.0  196.0   0.524    0.359   0.756  
2   47.0  52.0   59.0   0.441    0.375   0.830  
3    1.0   1.0    5.0   0.452    0.250   1.000  
4   34.0  36.0   47.0   0.612    0.533   0.658  
5   27.0  22.0   28.0   0.380    0.230   0.750  
6   13.0   6.0   11.0   0.333    0.270   0.556  
7  121.0  20.0  153.0   0.493    0.317   0.837  
8    0.0   0.0    0.0   0.000    0.000   0.000  
9   28.0   3.0   43.0   0.391    0.363   0.859

**Q13: What columns represent a player's scoring ability?**

A: pts, reb, ast, stl, blk, tov, fg_pct, fg3_pct, ft_pct. (Maybe, I literally have no clue what sports stuff means.)

### Task 5.2: High Scorers

Find all players who averaged 20+ points per game in 2021-22.

**Hint:** Filter WHERE pts >= 20

In [13]:
# TODO: Find players averaging 20+ PPG
query = """
SELECT player_id, season, team_id, gp, pts, reb, ast
FROM player_season_stats
WHERE season = '2021-22'
  AND pts >= 20
ORDER BY pts desc
"""

star_scorers = pd.read_sql(query, conn)
print(f"Found {len(star_scorers)} players averaging 20+ PPG")
display(star_scorers.head(20))

Found 513 players averaging 20+ PPG


player_id   season     team_id  gp     pts     reb    ast
0     1629027  2021-22  1610612737  76  2155.0   284.0  737.0
1      201942  2021-22  1610612741  76  2118.0   392.0  374.0
2      203954  2021-22  1610612755  68  2079.0   796.0  284.0
3     1628369  2021-22  1610612738  76  2046.0   609.0  334.0
4      203999  2021-22  1610612743  74  2004.0  1019.0  584.0
5      203507  2021-22  1610612749  67  2002.0   778.0  388.0
6     1629029  2021-22  1610612742  65  1847.0   593.0  568.0
7     1626164  2021-22  1610612756  68  1822.0   342.0  329.0
8     1626157  2021-22  1610612750  74  1818.0   727.0  269.0
9     1628378  2021-22  1610612762  67  1733.0   282.0  358.0
10       2544  2021-22  1610612747  56  1695.0   459.0  349.0
11     201142  2021-22  1610612751  55  1643.0   407.0  351.0
12     203897  2021-22  1610612741  67  1635.0   308.0  303.0
13     201939  2021-22  1610612744  64  1630.0   335.0  404.0
14    1628970  2021-22  1610612766  80  1613.0   559.0  300.0
15    1629630  2021-22  1610612763  57  1564.0   325.0  384.0
16    1627759  2021-22  1610612738  66  1559.0   404.0  231.0
17    1627783  2021-22  1610612761  68  1551.0   580.0  360.0
18    1630162  2021-22  1610612750  72  1533.0   343.0  275.0
19    1630163  2021-22  1610612766  75  1508.0   501.0  571.0

**Q14: How many players averaged 20+ PPG?**

A: 20

**Q15: Why might you want to filter by games played (gp) too?**

A: to gain more context on the effort put in.

### 📊 Task 5.3: Export Top Scorers to Excel

Export top scorers (20+ PPG, 40+ games) to Excel for analysis.

In [28]:
# TODO: Get top scorers with at least 40 games played
query = """
SELECT player_id, season, team_id, gp, min, pts, reb, ast, stl, blk, fg_pct, fg3_pct, ft_pct
FROM player_season_stats
WHERE gp >= 40
ORDER BY pts DESC
"""

top_scorers = pd.read_sql(query, conn)

# TODO: Export to Excel
high_scoring_all.to_excel(
    "top_scorers.xlsx",
    index=False,
    sheet_name="Top Scoring Player"
)

print(f"✅ Exported {len(top_scorers)} top scorers to 'top_scorers.xlsx'")
display(top_scorers.head(10))

✅ Exported 1419 top scorers to 'top_scorers.xlsx'


player_id   season     team_id  gp          min     pts    reb    ast  \
0    1628983  2024-25  1610612760  76  2597.630000  2484.0  379.0  486.0   
1    1629029  2023-24  1610612742  70  2624.040000  2370.0  647.0  686.0   
2    1628983  2023-24  1610612760  75  2552.686667  2254.0  415.0  465.0   
3    1628369  2022-23  1610612738  74  2732.196667  2225.0  649.0  342.0   
4     203507  2023-24  1610612749  73  2567.243333  2222.0  841.0  476.0   
5    1628973  2023-24  1610612752  77  2726.303333  2212.0  278.0  519.0   
6     203954  2022-23  1610612755  66  2284.106667  2183.0  670.0  274.0   
7    1630162  2024-25  1610612750  79  2871.058333  2177.0  450.0  359.0   
8    1629027  2021-22  1610612737  76  2651.891667  2155.0  284.0  737.0   
9    1629029  2022-23  1610612742  66  2390.471667  2138.0  569.0  529.0   

     stl    blk  fg_pct  fg3_pct  ft_pct  
0  131.0   77.0   0.519    0.375   0.898  
1   99.0   38.0   0.487    0.382   0.786  
2  150.0   67.0   0.535    0.353   0.874  
3   78.0   51.0   0.466    0.350   0.854  
4   87.0   79.0   0.611    0.274   0.657  
5   70.0   13.0   0.479    0.401   0.847  
6   66.0  112.0   0.548    0.330   0.857  
7   91.0   51.0   0.447    0.395   0.837  
8   72.0    7.0   0.460    0.382   0.904  
9   90.0   33.0   0.496    0.342   0.742

**Q16: How many top scorers were exported?**

A: 10

### Task 5.4: All-Around Players

Find players who contributed across multiple categories:
- 15+ points per game
- 5+ rebounds per game
- 5+ assists per game
- Played at least 40 games

**Hint:** You'll need 4 AND conditions!

In [17]:
# TODO: Find all-around players (15+ pts, 5+ reb, 5+ ast, 40+ games)
query = """
SELECT player_id, season, team_id, gp, pts, reb, ast
FROM player_season_stats
WHERE season = '2021-22'
  AND pts >= 15
  AND reb >= 5
  AND ast >= 5
  AND gp >= 40
ORDER BY gp DESC
"""

all_around = pd.read_sql(query, conn)
print(f"All-around players: {len(all_around)} found")
display(all_around)

All-around players: 357 found


player_id   season     team_id  gp     pts    reb    ast
0       203939  2021-22  1610612742  82   717.0  404.0   97.0
1      1626172  2021-22  1610612744  82   491.0  596.0  165.0
2      1628969  2021-22  1610612756  82  1162.0  347.0  185.0
3      1630166  2021-22  1610612764  82   685.0  425.0  167.0
4      1630180  2021-22  1610612765  82  1321.0  441.0  233.0
..         ...      ...         ...  ..     ...    ...    ...
352     203078  2021-22  1610612764  40   928.0  189.0  265.0
353     203482  2021-22  1610612765  40   365.0  177.0  111.0
354    1627823  2021-22  1610612762  40   131.0  100.0   20.0
355    1629744  2021-22  1610612741  40   161.0   50.0   20.0
356    1630540  2021-22  1610612752  40    89.0   42.0   41.0

[357 rows x 7 columns]

**Q17: Why use multiple AND conditions instead of OR?**

A: Because or would allow someone with less than 40 games to be included if ANY of the other conditions apply, and vice versa for the pther conditions.

---
## Part 6: Win/Loss Records

Let's create a win/loss dataset that we can analyze in Excel!

### Task 6.1: Get All 2021-22 Game Results

Get all games from 2021-22 with key stats.

In [19]:
# TODO: Get all 2021-22 games with important stats
query = """
SELECT season, game_date, team_id, matchup, wl, pts, fgm, fga, fg3m, ftm, reb, ast, stl, blk, tov
FROM team_game_stats
WHERE season = '2021-22'
ORDER BY season DESC
"""

all_games = pd.read_sql(query, conn)
print(f"Total games: {len(all_games)}")
display(all_games.head(10))

Total games: 2460


season   game_date     team_id      matchup wl  pts  fgm  fga  fg3m  ftm  \
0  2021-22  2021-10-19  1610612749  MIL vs. BKN  W  127   48  105    17   14   
1  2021-22  2021-10-19  1610612751    BKN @ MIL  L  104   37   84    17   13   
2  2021-22  2021-10-19  1610612744    GSW @ LAL  W  121   41   93    14   25   
3  2021-22  2021-10-19  1610612747  LAL vs. GSW  L  114   45   95    15    9   
4  2021-22  2021-10-20  1610612754    IND @ CHA  L  122   42   90    17   21   
5  2021-22  2021-10-20  1610612766  CHA vs. IND  W  123   46  107    13   18   
6  2021-22  2021-10-20  1610612741    CHI @ DET  W   94   37   86     7   13   
7  2021-22  2021-10-20  1610612765  DET vs. CHI  L   88   36   90     6   10   
8  2021-22  2021-10-20  1610612738    BOS @ NYK  L  134   48  117    21   17   
9  2021-22  2021-10-20  1610612752  NYK vs. BOS  W  138   51  105    17   19   

   reb  ast  stl  blk  tov  
0   54   25    8    9    8  
1   44   19    3    9   13  
2   50   30    9    2   17  
3   45   21    7    4   18  
4   51   29    2   10   17  
5   46   29    9    5    8  
6   48   18    8    5   17  
7   47   17    7    5   17  
8   56   34   13    9   18  
9   55   27    9   10   19

**Q18: How many total game records are there?**

A: 2460, 10 displayed

### 📊 Task 6.2: Export Win/Loss Records to Excel

Export this data - we'll use it for Excel analysis!

In [29]:
# TODO: Export all_games to Excel
all_games.to_excel(
    "win_loss_records.xlsx",
    index=False,
    sheet_name="All Games Score"
)

print(f"✅ Exported {len(all_games)} game records to 'win_loss_records.xlsx'")
print("\nWin/Loss distribution:")
print(all_games['wl'].value_counts())

✅ Exported 2460 game records to 'win_loss_records.xlsx'

Win/Loss distribution:
wl
W    1230
L    1230
Name: count, dtype: int64


---
## Part 7: Creating ML-Ready Dataset

Now let's create the dataset we'll use to train a machine learning model!

### Task 7.1: Win Prediction Dataset

Create a dataset for predicting wins. Include features that might predict winning:
- Points scored
- Field goals made/attempted
- Three-pointers made
- Free throws made
- Rebounds
- Assists
- Steals
- Blocks
- Turnovers
- Win/Loss outcome

In [23]:
# TODO: Select all the columns listed above from team_game_stats for 2021-22
query = """
SELECT 
    pts,
    fgm,
    fg3m,
    ftm,
    reb,
    ast,
    stl,
    blk,
    tov,
    wl
FROM team_game_stats
WHERE season = '2021-22'
"""

ml_dataset = pd.read_sql(query, conn)
print(f"ML Dataset created: {ml_dataset.shape[0]} games, {ml_dataset.shape[1]} features")
print("\nFirst 10 rows:")
display(ml_dataset.head(10))

print("\nDataset info:")
print(ml_dataset.info())

print("\nTarget variable distribution:")
print(ml_dataset['wl'].value_counts())

ML Dataset created: 2460 games, 10 features

First 10 rows:


pts  fgm  fg3m  ftm  reb  ast  stl  blk  tov wl
0  127   48    17   14   54   25    8    9    8  W
1  104   37    17   13   44   19    3    9   13  L
2  121   41    14   25   50   30    9    2   17  W
3  114   45    15    9   45   21    7    4   18  L
4  122   42    17   21   51   29    2   10   17  L
5  123   46    13   18   46   29    9    5    8  W
6   94   37     7   13   48   18    8    5   17  W
7   88   36     6   10   47   17    7    5   17  L
8  134   48    21   17   56   34   13    9   18  L
9  138   51    17   19   55   27    9   10   19  W


Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   pts     2460 non-null   int64 
 1   fgm     2460 non-null   int64 
 2   fg3m    2460 non-null   int64 
 3   ftm     2460 non-null   int64 
 4   reb     2460 non-null   int64 
 5   ast     2460 non-null   int64 
 6   stl     2460 non-null   int64 
 7   blk     2460 non-null   int64 
 8   tov     2460 non-null   int64 
 9   wl      2460 non-null   object
dtypes: int64(9), object(1)
memory usage: 192.3+ KB
None

Target variable distribution:
wl
W    1230
L    1230
Name: count, dtype: int64


**Q19: Is this dataset balanced (roughly equal wins and losses)?**

A: Yes

**Q20: Why is it important to include turnovers (tov) as a feature?**

A: I don't know, because I don't know what a turnover is. Something about possesing the ball and losing it? Which I'm going to guess is important because it shows a struggle? or something?

### Task 7.2: Save ML Dataset

Save this dataset both as CSV (for Python ML) and Excel (for analysis).

In [31]:
# TODO: Save as CSV for machine learning
ml_dataset.to_csv('nba_win_prediction.xlsx', index=False)

# TODO: Also save as Excel for analysis
ml_dataset.to_excel("Nba Machine Learning.xlsx", index=False, sheet_name='Nba Machine learning set')

print("✅ ML Dataset saved:")
print("   - nba_win_prediction.csv (for Python ML)")
print("   - nba_win_prediction.xlsx (for Excel analysis)")
print(f"   - {len(ml_dataset)} rows")
print(f"   - {len(ml_dataset.columns)} columns")

✅ ML Dataset saved:
   - nba_win_prediction.csv (for Python ML)
   - nba_win_prediction.xlsx (for Excel analysis)
   - 2460 rows
   - 10 columns


---
## Part 8: Summary of Excel Files Created

Let's verify all the Excel files we've created for later analysis.

In [32]:
import os

excel_files = [
    'high_scoring_games.xlsx',
    'top_scorers.xlsx',
    'win_loss_records.xlsx',
    'nba_win_prediction.xlsx'
]

print("📊 Excel Files Created for MOS Practice:\n")
for i, file in enumerate(excel_files, 1):
    if os.path.exists(file):
        size = os.path.getsize(file) / 1024  # Size in KB
        print(f"{i}. ✅ {file} ({size:.1f} KB)")
    else:
        print(f"{i}. ❌ {file} (NOT FOUND - did you complete that task?)")

print("\n💡 Next Steps:")
print("- Open these files in Excel")
print("- Practice formulas (SUM, AVERAGE, IF, COUNTIF)")
print("- Create charts and visualizations")
print("- Apply conditional formatting")
print("- Build PivotTables")

📊 Excel Files Created for MOS Practice:

1. ✅ high_scoring_games.xlsx (191.6 KB)
2. ✅ top_scorers.xlsx (191.6 KB)
3. ✅ win_loss_records.xlsx (163.8 KB)
4. ✅ nba_win_prediction.xlsx (68.9 KB)

💡 Next Steps:
- Open these files in Excel
- Practice formulas (SUM, AVERAGE, IF, COUNTIF)
- Create charts and visualizations
- Apply conditional formatting
- Build PivotTables


---
## Part 9: Reflection Questions

**Q21: How is using SQL different from loading an entire CSV file?**

A: SQL is a coding language.

**Q22: What's the connection between SQL and machine learning?**

A: SQL helps manage databases while ML uses databases.

**Q23: When would you use WHERE vs. filtering in pandas?**

A: When navigating databases like this NBA one.

**Q24: What features would you add to improve a win prediction model?**

A: I'm telling you I seriously have no idea how sports worjs

**Q25: Why export to Excel instead of just working in Python?**

A: 

**Q26: How would SQL help if you had 50 million game records instead of 2,000?**

A: 

---
## Cleanup

In [ ]:
# Always close your database connection when done
conn.close()
print("✅ Database connection closed")

---
## Summary

**You practiced:**
- ✅ Connecting to databases
- ✅ Exploring table structures
- ✅ SELECT queries (choosing features)
- ✅ WHERE filters (selecting training examples)
- ✅ ORDER BY (finding extremes and patterns)
- ✅ LIMIT (controlling result size)
- ✅ Multiple conditions (AND)
- ✅ Loading SQL results into pandas
- ✅ Exporting data to Excel for analysis

**Excel Files Created:**
1. high_scoring_games.xlsx
2. top_scorers.xlsx
3. win_loss_records.xlsx
4. nba_win_prediction.xlsx

**Key Takeaway:** SQL is how you extract and prepare data for machine learning in the real world!

---

## Next Steps

**In the next lessons, you'll:**
- Open these Excel files
- Practice MOS certification skills (formulas, charts, formatting)
- Learn JOIN queries (combining tables)
- Learn GROUP BY (aggregating data)
- Build a win prediction model with your SQL data!

**Great work!**